File: libs/contracts/events.ts
// ──────────────────────────────────────────────────────────────
// Propeller Event Contracts — shared between Go & TS
// ──────────────────────────────────────────────────────────────
// These event type strings are the canonical names used on
// flo streams. Go services publish/consume these; TS apps
// read these event shapes for UI/API responses.
// ──────────────────────────────────────────────────────────────

// ── Event type constants (keep in sync with Go contracts/events.go) ──

export const EventTypes = {
  // Webhook — raw inbound events (single stream, routed by flo pipeline)
  WebhookReceived: 'webhook.received',

  // Payment collection (Paystack PWT)
  PaymentReceived: 'payment.received',
  PaymentRejected: 'payment.rejected',
  PaymentCollected: 'payment.collected',

  // KYB (PayKKa onboarding)
  KybSubmitted: 'kyb.submitted',
  KybIdentityStarted: 'kyb.identity_started',
  KybIdentityPending: 'kyb.identity_pending',
  KybCompleted: 'kyb.completed',
  KybApproved: 'kyb.approved',
  KybRejected: 'kyb.rejected',
  KybManualReview: 'kyb.manual_review',

  // KYC (Dojah)
  KycCompleted: 'kyc.completed',

  // Payout (Globalstack off-ramp)
  PayoutRequested: 'payout.requested',
  PayoutDelivered: 'payout.delivered',
  PayoutSettled: 'payout.settled',
  PayoutFailed: 'payout.failed',

  // Business lifecycle
  BusinessActivated: 'business.activated',
  BusinessSuspended: 'business.suspended',
  BusinessClosed: 'business.closed',

  // Ledger
  LedgerCredited: 'ledger.credited',
  LedgerDebited: 'ledger.debited',
} as const;

export type EventType = (typeof EventTypes)[keyof typeof EventTypes];

// ── Action names (flo action handlers) ──

export const ActionNames = {
  ReceivePayment: 'receive-payment',
  RejectPayment: 'reject-payment',
  SubmitKyb: 'submit-kyb',
  CompleteKyb: 'complete-kyb',
  CompleteKyc: 'complete-kyc',
  RoutePayout: 'route-payout',
  DeliverPayout: 'deliver-payout',
  SettlePayout: 'settle-payout',
  FailPayout: 'fail-payout',
  ActivateBusiness: 'activate-business',
  DebitLedger: 'debit-ledger',
  CreditLedger: 'credit-ledger',
} as const;

export type ActionName = (typeof ActionNames)[keyof typeof ActionNames];

// ── Stream names ──

export const StreamNames = {
  RawWebhooks: 'raw-webhooks',
  PaymentEvents: 'payment-events',
  KybEvents: 'kyb-events',
  KycEvents: 'kyc-events',
  PayoutEvents: 'payout-events',
  BusinessEvents: 'business-events',
  LedgerEvents: 'ledger-events',
} as const;

// ── Event payload shapes ──

export interface WebhookRecord {
  source: 'paystack' | 'paykka' | 'dojah' | 'globalstack';
  event: string;
  payload: Record<string, unknown>;
  headers: Record<string, string>;
  receivedAt: string;
  idempotencyKey: string;
}

export interface PaymentReceivedEvent {
  provider: 'paystack';
  reference: string;
  amount: number; // kobo
  currency: string;
  senderName: string;
  senderBankAccount: string;
  senderCountry: string;
  paidAt: string;
  paymentRequestId: string;
}

export interface PaymentRejectedEvent {
  provider: 'paystack';
  reference: string;
  amount: string;
  message: string;
  messageType: string;
}

export interface KybCompletedEvent {
  provider: 'paykka';
  merchId: string;
  status: 'APPROVED' | 'REJECTED' | 'SUPPLEMENT';
  message?: string;
}

export interface KycCompletedEvent {
  provider: 'dojah';
  referenceId: string;
  status: 'approved' | 'declined' | 'pending';
  fullName?: string;
  dob?: string;
  country?: string;
  docType?: string;
  livenessScore?: number;
}

export interface PayoutSettledEvent {
  provider: 'globalstack';
  orderId: string;
  targetTxHash?: string;
  targetAmount?: number;
  targetCurrency?: string;
}

export interface LedgerEntry {
  accountId: bigint;
  accountType: string;
  amount: bigint;
  direction: 'debit' | 'credit';
  transferId: bigint;
  businessId: string;
}

File: libs/go-events/events.go
// Package events defines the canonical event types, action names,
// and stream names shared across Go services and TS apps.
//
// Keep in sync with libs/contracts/events.ts
package events

// ── Event types ──

const (
	// Webhook
	EventWebhookReceived = "webhook.received"

	// Payment collection (Paystack PWT)
	EventPaymentReceived  = "payment.received"
	EventPaymentRejected  = "payment.rejected"
	EventPaymentCollected = "payment.collected"

	// KYB (PayKKa risk assessment)
	EventKybSubmitted    = "kyb.submitted"
	EventKybCompleted    = "kyb.completed"
	EventKybApproved     = "kyb.approved"
	EventKybRejected     = "kyb.rejected"
	EventKybManualReview = "kyb.manual_review"

	// KYC (Dojah)
	EventKycCompleted = "kyc.completed"

	// Payout
	EventPayoutRequested = "payout.requested"
	EventPayoutDelivered = "payout.delivered"
	EventPayoutSettled   = "payout.settled"
	EventPayoutFailed    = "payout.failed"

	// Business lifecycle
	EventBusinessActivated    = "business.activated"
	EventBusinessSuspended    = "business.suspended"
	EventBusinessClosed       = "business.closed"
	EventBusinessKycSubmitted = "business.kyc-submitted"

	// Ledger
	EventLedgerCredited = "ledger.credited"
	EventLedgerDebited  = "ledger.debited"
)

// ── Action names ──

const (
	ActionReceivePayment   = "receive-payment"
	ActionRejectPayment    = "reject-payment"
	ActionSubmitKyb        = "submit-kyb"
	ActionCompleteKyb      = "complete-kyb"
	ActionCompleteKyc      = "complete-kyc"
	ActionRoutePayout      = "route-payout"
	ActionDeliverPayout    = "deliver-payout"
	ActionSettlePayout     = "settle-payout"
	ActionFailPayout       = "fail-payout"
	ActionActivateBusiness = "activate-business"
	ActionDebitLedger      = "debit-ledger"
	ActionCreditLedger     = "credit-ledger"
)

// ── Stream names ──

const (
	StreamRawWebhooks        = "raw-webhooks"
	StreamPaymentEvents      = "payment-events"
	StreamKybEvents          = "kyb-events"
	StreamKycEvents          = "kyc-events"
	StreamPayoutEvents       = "payout-events"
	StreamBusinessEvents     = "business-events"
	StreamNotificationEvents = "notification-events"
	StreamLedgerEvents       = "ledger-events"
)

// ── Payload types ──

type WebhookRecord struct {
	Source         string                 `json:"source"` // paystack | paykka | dojah | globalstack
	Event          string                 `json:"event"`  // original provider event name
	Payload        map[string]interface{} `json:"payload"`
	Headers        map[string]string      `json:"headers"`
	ReceivedAt     string                 `json:"receivedAt"`
	IdempotencyKey string                 `json:"idempotencyKey"`
}

type PaymentReceivedEvent struct {
	Provider          string `json:"provider"`
	Reference         string `json:"reference"`
	Amount            int64  `json:"amount"` // kobo
	Currency          string `json:"currency"`
	SenderName        string `json:"senderName"`
	SenderBankAccount string `json:"senderBankAccount"`
	SenderCountry     string `json:"senderCountry"`
	PaidAt            string `json:"paidAt"`
	PaymentRequestID  string `json:"paymentRequestId"`
}

type PaymentRejectedEvent struct {
	Provider    string `json:"provider"`
	Reference   string `json:"reference"`
	Amount      string `json:"amount"`
	Message     string `json:"message"`
	MessageType string `json:"messageType"`
}

type KybSubmittedEvent struct {
	BusinessID string `json:"businessId"`
	RequestID  string `json:"requestId"`
	MerchID    string `json:"merchId"`
}

type BusinessSubmissionReadyEvent struct {
	BusinessID string `json:"businessId"`
	RequestID  string `json:"requestId"`
}

type KybCompletedEvent struct {
	Provider  string `json:"provider"`
	MerchID   string `json:"merchId"`
	RequestID string `json:"requestId,omitempty"`
	Status    string `json:"status"`              // INIT | WAIT | PASS | REFUSED | AUTH_FAIL | REJECTED
	RiskLevel string `json:"riskLevel,omitempty"` // LOW | MIDDLE | HIGH
	Message   string `json:"message,omitempty"`
}

type BusinessActivatedEvent struct {
	BusinessID string `json:"businessId"`
	MerchID    string `json:"merchId"`
	Provider   string `json:"provider"`
}

type KycCompletedEvent struct {
	Provider      string `json:"provider"`
	ReferenceID   string `json:"referenceId"`
	Status        string `json:"status"`
	FullName      string `json:"fullName,omitempty"`
	DOB           string `json:"dob,omitempty"`
	Country       string `json:"country,omitempty"`
	DocType       string `json:"docType,omitempty"`
	LivenessScore int    `json:"livenessScore,omitempty"`
}

type PayoutSettledEvent struct {
	Provider       string `json:"provider"`
	OrderID        string `json:"orderId"`
	TargetTxHash   string `json:"targetTxHash,omitempty"`
	TargetAmount   int64  `json:"targetAmount,omitempty"`
	TargetCurrency string `json:"targetCurrency,omitempty"`
}
